### ISSUE TITLE
mixed precision training, hyperparameters and bleu diverge

### ISSUE THREAD

**Issue by LiweiPeng** [NVIDIA/OpenSeq2Seq#271]
I am using the Transformer model to train a NMT dataset. From the openseq2seq paper (https://devblogs.nvidia.com/mixed-precision-nlp-speech-openseq2seq/), when using mixed precision, no hyper parameters (like learning rate) are required to change. Can you confirm that?

In my training, when fp32 is used, the BLEU score converge well and can reach ~0.3400. However, with same hyper parameters, the mixed precision training got very bad BLEU curve. The BLEU score first go up to something like 0.3, then gradually go down to close to 0.01.

Can someone share some insights on what's wrong? Thanks.

System I used:
 - V100
 - tensorflow 1.11.0,
 - cuda 9.2
 - cudnn 7.3.1
 - Horovod 0.15.0

**Comment by rahul003** [comment]
Did you enable loss scaling? You might be running into gradient underflow. In some cases loss scaling is necessary to get similar accuracies.

**Comment by okuchaiev** [comment]
yep, @LiweiPeng  do you have
```
  "dtype": "mixed",
  "loss_scaling": "Backoff",
```
set? Also, which branch/tag of OpenSeq2Seq do you use?

**Comment by LiweiPeng** [comment]
I am not using OpenSeq2Seq. I borrowed the design from OpenSeq2Seq and reimplemented in my own project. I'll check the Backoff loss_scaling design. Thanks.

**Comment by okuchaiev** [comment]
@LiweiPeng In this case I recommend you take:
(1) Loss Scaling https://github.com/NVIDIA/OpenSeq2Seq/blob/master/open_seq2seq/optimizers/automatic_loss_scaler.py and
(2) MixedPrecision optimizer wrapper: https://github.com/NVIDIA/OpenSeq2Seq/blob/master/open_seq2seq/optimizers/mp_wrapper.py 

They are not dependent on anything else in OpenSeq2Seq so you should be able to easily use them in your project.

**Comment by okuchaiev** [comment]
closing since this isn't related to OpenSeq2Seq codebase
